### DISCUSSION TITLE
Cross Entropy Loss and loss of PyTorch Lightning does not matches

### DISCUSSION THREAD

**Discussion by ayush714** [Lightning-AI/pytorch-lightning#9159] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi, I am working on building a question and answering model using T5(Huggingface) with PyTorch Lightning Module and I am checking my loss and PyTorch Lightning Loss is not being matched.

```
class UQAFineTuneModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = T5ForConditionalGeneration.from_pretrained(
            "allenai/unifiedqa-t5-small", return_dict=True
        )
        self.model.train()
    def forward(
        self,
        source_text_input_ids,
        source_text_attention_mask,
        target_text_input_ids=None,
    ):
        output = self.model(
            input_ids=source_text_input_ids,
            attention_mask=source_text_attention_mask,
            labels=target_text_input_ids,
        )
        return output.loss, output.logits

    def training_step(self, batch, batch_idx):
        source_text_input_ids = batch["source_text_input_ids"]
        source_text_attention_mask = batch["source_text_attention_mask"]
        target_text_input_ids = batch["target_text_input_ids"]
        # labels_attention_mask = batch["target_text_attention_mask"]
        loss, outputs = self(
            source_text_input_ids, source_text_attention_mask, target_text_input_ids
        ) 
        loss_mine = None  
        output = self.model(
            input_ids=source_text_input_ids,
            attention_mask=source_text_attention_mask,
            labels=target_text_input_ids,
        ) 
        labels = batch["target_text_input_ids"].clone() 
        labels[labels == 0] = -100 
        if target_text_input_ids is not None:  
            loss_fct = CrossEntropyLoss(ignore_index=-100) 
            loss_mine = loss_fct(output.logits.view(-1, outputs.size(-1)), labels.view(-1)) 
            print(f"loss_Hugginface: {loss.item()}, loss_mine : {loss_mine.item()}") 
        self.log("train_loss", loss, prog_bar=True, logger=True)
        return {"loss": loss, "predictions": outputs}
```

![image](https://user-images.githubusercontent.com/81796368/131071640-14d4dade-c9ec-4c6a-8ec3-d04f0ed5411c.png)
You can see the above image, why loss is not same, help is very much needed, I asked the same question on HuggingFace but they told me to ask here, you can view that discussion here.

**Comment by tchaton** [comment, marked as answer]
Hey @ayush714,

I believe this is related to HF and you might get your answers by opening an issue on their repo directly.

Best,
T.C

**Reply by ayush714** [reply]
Hi @tchaton I asked but they told to ask here, here is the full response:- 
This seems a question that pertains to PyTorch Lightning and not Transformers. Better to ask the people over there, specifically how their loss function is implemented and - importantly - how they reshape the labels and outputs.

**Reply by tchaton** [reply]
ahahah :) Lightning doesn't do any manipulation on the labels / outputs. 
This is the responsibility of the user.
From your code, I can clearly see you are just trying to understand how does HF models compute their loss by replicating it manually within the Lightning training_step.

I would encourage you to dive within the HF transformers code base to where the loss is computed and compare inputs tensors to yours.
My guess is this line might be the reason `labels[labels == 0] = -100 `.

I hope you will find the answer :)

Best,
T.C
